{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "ddfe775a-8d8a-43d0-a25a-6de438f06cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/sc/arion/work/massen06/.conda/envs/qc_torch/bin/python3.9\n",
      "     active environment : qc_torch\n",
      "    active env location : /sc/arion/work/massen06/.conda/envs/qc_torch\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.executable)\n",
    "!conda info | grep 'active env'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1b222304-72ca-4d5b-b389-1f186d3576b8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import copy\n",
    "import scipy\n",
    "import shutil\n",
    "import logging\n",
    "import h5py\n",
    "import tables\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from datetime import datetime\n",
    "import scipy.sparse as sparse\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae9585e-5645-4b34-ab75-6bf4ff510030",
   "metadata": {},
   "source": [
    "# Mouse Living Brain - Data Consolidation\n",
    "\n",
    "Nicolas Masse, Donghoon Lee  \n",
    "July 2023  \n",
    "\n",
    "Assumes ambient mRNA has already been removed using ambient_removal.ipynb. This step will save the data in .h5 format in the base_dir (see below)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3b0ad4-711d-44a9-ac1e-8d042a908b24",
   "metadata": {},
   "source": [
    "### Paths for metadata spreadsheet, gene info, and target path for AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "5dacdda8-bd66-431a-b7e0-27742d02be31",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_dir = \"/sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/metadata\"\n",
    "base_dir = \"/sc/arion/work/massen06/living_brain1/ambient_removal\"\n",
    "save_dir = \"/sc/arion/work/massen06/living_brain1/anndata\"\n",
    "gene_info_fn = \"./data/MM_GRCm39_full.txt\" \n",
    "mouse_mitocarta_fn = \"./data/MouseMitoCarta3.0.csv\" "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc6be895-0d2b-406c-8abd-f031dc5cab91",
   "metadata": {},
   "source": [
    "### Extract metadata from Living_brain_snRNA_template.xlsx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "294579c9-6577-4fe9-b24b-60b01780d74d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Mouse sample Info',\n",
       " 'Human sample Info',\n",
       " 'Master Day 1',\n",
       " 'Evelyn',\n",
       " 'Clara',\n",
       " 'Mouse_cDNA_Libraries',\n",
       " 'Human_cDNA_libraries',\n",
       " 'HTOs',\n",
       " 'Pooling_rand',\n",
       " 'Pooling',\n",
       " 'Barcodes',\n",
       " 'Ab Barcodes']"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metadata_fn = os.path.join(metadata_dir, \"Living_brain_snRNA_template.xlsx\")\n",
    "metadata = pd.read_excel(metadata_fn, sheet_name=\"Mouse_cDNA_Libraries\")\n",
    "#metadata = pd.read_excel(metadata_fn)\n",
    "xl = pd.ExcelFile(metadata_fn)\n",
    "xl.sheet_names  # see all sheet names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "d12d1305-e8c7-441b-9920-40c98ad42c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_donor_information(data):\n",
    "    \"\"\"Extract donor name, PMI and living status\"\"\"\n",
    "    \n",
    "    donor_info = {\n",
    "        \"unique_id\": [],\n",
    "        \"SubID\" : [],\n",
    "        \"pmi\": [],\n",
    "        \"living\": [],\n",
    "    }\n",
    "\n",
    "    for d in data.values:\n",
    "        \n",
    "        # get donor name\n",
    "        i0 = d.find(\"_\")\n",
    "        donor_info[\"SubID\"].append(d[: i0])\n",
    "        \n",
    "        # get PMI\n",
    "        time = int(\"\".join([s for s in d[i0 :] if s.isdigit()]))\n",
    "        if \"hr\" in d[i0 :]:\n",
    "            time *= 60\n",
    "        donor_info[\"pmi\"].append(time)\n",
    "        \n",
    "        if \"PM\" in d or \"Post\" in d or time > 0:\n",
    "            donor_info[\"living\"].append(\"PM\")\n",
    "        else:\n",
    "            donor_info[\"living\"].append(\"Living\")\n",
    "            \n",
    "        un_id = donor_info[\"SubID\"][-1] + \"_\" +  donor_info[\"living\"][-1]\n",
    "        donor_info[\"unique_id\"].append(un_id)\n",
    "        \n",
    "    return donor_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "bb835560-fcd8-4461-9671-986b63d93c0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "     unique_id SubID  pmi  living step_2_date  cdna_amp_cycle  input_ng  \\\n",
      "0   WTA_Living   WTA    0  Living    06022023              11     139.6   \n",
      "1       WTA_PM   WTA    0      PM    06022023              11     174.0   \n",
      "2   912_Living   912    0  Living    06022023              11      40.8   \n",
      "3       912_PM   912    5      PM    06022023              11     146.8   \n",
      "4   WTF_Living   WTF    0  Living    06022023              11      31.4   \n",
      "5       WTF_PM   WTF   60      PM    06022023              11     140.4   \n",
      "6   WTB_Living   WTB    0  Living    06062023              12      28.0   \n",
      "7       WTB_PM   WTB    0      PM    06062023              12      40.0   \n",
      "8   913_Living   913    0  Living    06062023              12      21.4   \n",
      "9       913_PM   913    5      PM    06062023              12      47.6   \n",
      "10  WTE_Living   WTE    0  Living    06062023              12      16.0   \n",
      "11      WTE_PM   WTE  360      PM    06062023              12      50.0   \n",
      "12  WTC_Living   WTC    0  Living    06072023              12     460.0   \n",
      "13      WTC_PM   WTC    0      PM    06072023              12     384.0   \n",
      "14  WTG_Living   WTG    0  Living    06072023              12     164.0   \n",
      "15      WTG_PM   WTG   60      PM    06072023              12     404.0   \n",
      "16  WTD_Living   WTD    0  Living    06072023              12      34.2   \n",
      "17      WTD_PM   WTD  360      PM    06072023              12     296.0   \n",
      "\n",
      "    pcr_cycle  \n",
      "0          12  \n",
      "1          12  \n",
      "2          13  \n",
      "3          12  \n",
      "4          13  \n",
      "5          12  \n",
      "6          14  \n",
      "7          13  \n",
      "8          14  \n",
      "9          13  \n",
      "10         14  \n",
      "11         13  \n",
      "12         10  \n",
      "13         10  \n",
      "14         12  \n",
      "15         10  \n",
      "16         13  \n",
      "17         11  \n"
     ]
    }
   ],
   "source": [
    "for k in metadata.keys():\n",
    "    \n",
    "    column_name = metadata[k].iloc[0]\n",
    "\n",
    "    if column_name == \"unique sample ID\":\n",
    "        donor_info = extract_donor_information(metadata[k].iloc[1 :])\n",
    "        \n",
    "    elif column_name == \"Step 2 date\":\n",
    "        dates = metadata[k].iloc[1 :]\n",
    "        dates = [d.strftime(\"%m%d%Y\") for d in dates]\n",
    "        donor_info[\"step_2_date\"] = dates\n",
    "        \n",
    "    elif column_name == \"cDNA Amp cycle #s\":\n",
    "        cycles = metadata[k].iloc[1 :]\n",
    "        cycles = [int(c) for c in cycles]\n",
    "        donor_info[\"cdna_amp_cycle\"] = cycles\n",
    "        \n",
    "    elif column_name == \"input (ng)\":\n",
    "        inputs = metadata[k].iloc[1 :]\n",
    "        inputs = [round(float(i), 1) for i in inputs]\n",
    "        donor_info[\"input_ng\"] = inputs\n",
    "        \n",
    "    elif column_name == \"Indexing PCR cycle #\":\n",
    "        cycle = metadata[k].iloc[1 :]\n",
    "        cycle = [int(c) for c in cycle]\n",
    "        donor_info[\"pcr_cycle\"] = cycle\n",
    "        \n",
    "donor_info = pd.DataFrame(data = donor_info)\n",
    "\n",
    "print(donor_info)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e8bc0c-0324-4684-8ac7-37ab18f012e0",
   "metadata": {},
   "source": [
    "### Load gene info from Biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "efc2a921-f16c-48c2-8943-7107d0a513d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_75251/2442110492.py:1: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gene_info = pd.read_csv(gene_info_fn, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>ENSMUSG00000064336.1</td>\n",
       "      <td>ENSMUST00000082387</td>\n",
       "      <td>ENSMUST00000082387.1</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA phenylalanine [So...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>ENSMUSG00000064337.1</td>\n",
       "      <td>ENSMUST00000082388</td>\n",
       "      <td>ENSMUST00000082388.1</td>\n",
       "      <td>70</td>\n",
       "      <td>1024</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:MGI S...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>ENSMUSG00000064338.1</td>\n",
       "      <td>ENSMUST00000082389</td>\n",
       "      <td>ENSMUST00000082389.1</td>\n",
       "      <td>1025</td>\n",
       "      <td>1093</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA valine [Source:MG...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>ENSMUSG00000064339.1</td>\n",
       "      <td>ENSMUST00000082390</td>\n",
       "      <td>ENSMUST00000082390.1</td>\n",
       "      <td>1094</td>\n",
       "      <td>2675</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:MGI S...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>ENSMUSG00000064340.1</td>\n",
       "      <td>ENSMUST00000082391</td>\n",
       "      <td>ENSMUST00000082391.1</td>\n",
       "      <td>2676</td>\n",
       "      <td>2750</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA leucine 1 [Source...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56477</th>\n",
       "      <td>149518</td>\n",
       "      <td>ENSMUSG00000054766</td>\n",
       "      <td>ENSMUSG00000054766.14</td>\n",
       "      <td>ENSMUST00000134364</td>\n",
       "      <td>ENSMUST00000134364.8</td>\n",
       "      <td>29947390</td>\n",
       "      <td>29962589</td>\n",
       "      <td>Set</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SET nuclear oncogene [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56478</th>\n",
       "      <td>149525</td>\n",
       "      <td>ENSMUSG00000026785</td>\n",
       "      <td>ENSMUSG00000026785.15</td>\n",
       "      <td>ENSMUST00000125346</td>\n",
       "      <td>ENSMUST00000125346.8</td>\n",
       "      <td>29967696</td>\n",
       "      <td>29981034</td>\n",
       "      <td>Pkn3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein kinase N3 [Source:MGI Symbol;Acc:MGI:2...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56479</th>\n",
       "      <td>149533</td>\n",
       "      <td>ENSMUSG00000015335</td>\n",
       "      <td>ENSMUSG00000015335.17</td>\n",
       "      <td>ENSMUST00000148717</td>\n",
       "      <td>ENSMUST00000148717.8</td>\n",
       "      <td>29980956</td>\n",
       "      <td>29983660</td>\n",
       "      <td>Zdhhc12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>zinc finger, DHHC domain containing 12 [Source...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56480</th>\n",
       "      <td>149537</td>\n",
       "      <td>ENSMUSG00000039686</td>\n",
       "      <td>ENSMUSG00000039686.15</td>\n",
       "      <td>ENSMUST00000044751</td>\n",
       "      <td>ENSMUST00000044751.14</td>\n",
       "      <td>29987295</td>\n",
       "      <td>30014597</td>\n",
       "      <td>Zer1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>zyg-11 related, cell cycle regulator [Source:M...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56481</th>\n",
       "      <td>149546</td>\n",
       "      <td>ENSMUSG00000100747</td>\n",
       "      <td>ENSMUSG00000100747.6</td>\n",
       "      <td>ENSMUST00000187343</td>\n",
       "      <td>ENSMUST00000187343.2</td>\n",
       "      <td>30126807</td>\n",
       "      <td>30128235</td>\n",
       "      <td>1700084E18Rik</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RIKEN cDNA 1700084E18 gene [Source:MGI Symbol;...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56482 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index      Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0           0  ENSMUSG00000064336   ENSMUSG00000064336.1   ENSMUST00000082387   \n",
       "1           1  ENSMUSG00000064337   ENSMUSG00000064337.1   ENSMUST00000082388   \n",
       "2           2  ENSMUSG00000064338   ENSMUSG00000064338.1   ENSMUST00000082389   \n",
       "3           3  ENSMUSG00000064339   ENSMUSG00000064339.1   ENSMUST00000082390   \n",
       "4           4  ENSMUSG00000064340   ENSMUSG00000064340.1   ENSMUST00000082391   \n",
       "...       ...                 ...                    ...                  ...   \n",
       "56477  149518  ENSMUSG00000054766  ENSMUSG00000054766.14   ENSMUST00000134364   \n",
       "56478  149525  ENSMUSG00000026785  ENSMUSG00000026785.15   ENSMUST00000125346   \n",
       "56479  149533  ENSMUSG00000015335  ENSMUSG00000015335.17   ENSMUST00000148717   \n",
       "56480  149537  ENSMUSG00000039686  ENSMUSG00000039686.15   ENSMUST00000044751   \n",
       "56481  149546  ENSMUSG00000100747   ENSMUSG00000100747.6   ENSMUST00000187343   \n",
       "\n",
       "      Transcript stable ID version  Gene start (bp)  Gene end (bp)  \\\n",
       "0             ENSMUST00000082387.1                1             68   \n",
       "1             ENSMUST00000082388.1               70           1024   \n",
       "2             ENSMUST00000082389.1             1025           1093   \n",
       "3             ENSMUST00000082390.1             1094           2675   \n",
       "4             ENSMUST00000082391.1             2676           2750   \n",
       "...                            ...              ...            ...   \n",
       "56477         ENSMUST00000134364.8         29947390       29962589   \n",
       "56478         ENSMUST00000125346.8         29967696       29981034   \n",
       "56479         ENSMUST00000148717.8         29980956       29983660   \n",
       "56480        ENSMUST00000044751.14         29987295       30014597   \n",
       "56481         ENSMUST00000187343.2         30126807       30128235   \n",
       "\n",
       "           Gene name       Gene type  \\\n",
       "0              mt-Tf         Mt_tRNA   \n",
       "1            mt-Rnr1         Mt_rRNA   \n",
       "2              mt-Tv         Mt_tRNA   \n",
       "3            mt-Rnr2         Mt_rRNA   \n",
       "4             mt-Tl1         Mt_tRNA   \n",
       "...              ...             ...   \n",
       "56477            Set  protein_coding   \n",
       "56478           Pkn3  protein_coding   \n",
       "56479        Zdhhc12  protein_coding   \n",
       "56480           Zer1  protein_coding   \n",
       "56481  1700084E18Rik          lncRNA   \n",
       "\n",
       "                                        Gene description  \\\n",
       "0      mitochondrially encoded tRNA phenylalanine [So...   \n",
       "1      mitochondrially encoded 12S rRNA [Source:MGI S...   \n",
       "2      mitochondrially encoded tRNA valine [Source:MG...   \n",
       "3      mitochondrially encoded 16S rRNA [Source:MGI S...   \n",
       "4      mitochondrially encoded tRNA leucine 1 [Source...   \n",
       "...                                                  ...   \n",
       "56477  SET nuclear oncogene [Source:MGI Symbol;Acc:MG...   \n",
       "56478  protein kinase N3 [Source:MGI Symbol;Acc:MGI:2...   \n",
       "56479  zinc finger, DHHC domain containing 12 [Source...   \n",
       "56480  zyg-11 related, cell cycle regulator [Source:M...   \n",
       "56481  RIKEN cDNA 1700084E18 gene [Source:MGI Symbol;...   \n",
       "\n",
       "      Chromosome/scaffold name  \n",
       "0                           MT  \n",
       "1                           MT  \n",
       "2                           MT  \n",
       "3                           MT  \n",
       "4                           MT  \n",
       "...                        ...  \n",
       "56477                        2  \n",
       "56478                        2  \n",
       "56479                        2  \n",
       "56480                        2  \n",
       "56481                        2  \n",
       "\n",
       "[56482 rows x 11 columns]"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_csv(gene_info_fn, sep='\\t')\n",
    "gene_info.drop_duplicates(subset = \"Gene name\", inplace=True)\n",
    "gene_info.reset_index(inplace=True)\n",
    "gene_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "5cb1cadf-a7e6-48e9-b744-dbf282a05c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def link_gene_info(adata, gene_info, verbose=False):\n",
    "    \"\"\"Add gene info to AnnData\"\"\"\n",
    "    \n",
    "    adata.var = adata.var.merge(gene_info, left_on=\"gene_ids\", right_on='Gene stable ID', how=\"outer\")\n",
    "\n",
    "    good_genes_idx = [i for i, name in enumerate(adata.var[\"Gene name\"].values) if isinstance(name, str)]\n",
    "       \n",
    "    new_adata = ad.AnnData(adata.X[:, good_genes_idx])\n",
    "    new_adata.obs = adata.obs\n",
    "    new_adata.var[\"gene_name\"] = [v for n, v in enumerate(adata.var[\"Gene name\"].values) if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_type\"] = [v for n, v in enumerate(adata.var[\"Gene type\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_id\"] = [v for n, v in enumerate(adata.var[\"Gene stable ID\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_chrom\"] = [v for n, v in enumerate(adata.var[\"Chromosome/scaffold name\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_start\"] = [v for n, v in enumerate(adata.var[\"Gene start (bp)\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_end\"] = [v for n, v in enumerate(adata.var[\"Gene end (bp)\"].values)  if n in good_genes_idx]\n",
    "    \n",
    "    print(f\"Variables added. Number of genes remaining: {new_adata.X.shape[1]}\")\n",
    "    \n",
    "    return new_adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "ee0e1aef-af94-43b1-8583-d52dd5d0d01a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readCellbenderH5(filename):\n",
    "    \n",
    "    f = h5py.File(filename, 'r')\n",
    "    mat = f['matrix']\n",
    "    cols = ['barcodes','latent_cell_probability','latent_RT_efficiency']\n",
    "    #cols = [\"barcodes\"]\n",
    "    obsdict = {x:mat[x] for x in cols}\n",
    "    \n",
    "    adata=sc.AnnData(X=scipy.sparse.csr_matrix(\n",
    "        (\n",
    "            mat['data'][:], \n",
    "            mat['indices'][:], \n",
    "            mat['indptr'][:]\n",
    "        ), \n",
    "        shape=(mat['shape'][1],mat['shape'][0])),\n",
    "        var=pd.DataFrame(dict(mat['features'])),\n",
    "        obs=pd.DataFrame(obsdict),\n",
    "        #uns={'test_elbo':list(mat['test_elbo']),'test_epoch':list(mat['test_epoch'])}\n",
    "    )\n",
    "    for k in adata.var.keys():\n",
    "        adata.var[k]=[x.decode('ascii') for x in adata.var[k]]\n",
    "        \n",
    "    adata.var.index = [x for x in adata.var['name']]\n",
    "    adata.obs.barcodes = [str(x.decode('ascii')) for x in mat['barcodes']]\n",
    "    adata.obs.index = adata.obs.barcodes\n",
    "    #adata.obs.index = [x.decode('ascii') for x in mat['barcodes']]\n",
    "    #adata.obs.reset_index(adata.obs.barcodes)\n",
    "    \n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "1e224fd8-69e7-415e-a000-9d745ab414de",
   "metadata": {},
   "outputs": [],
   "source": [
    "def readCellbenderH5(filename, suffix):\n",
    "    \n",
    "    f = h5py.File(filename, 'r')\n",
    "    mat = f['matrix']\n",
    "    \n",
    "    x = scipy.sparse.csr_matrix(\n",
    "        (\n",
    "            mat['data'][:], \n",
    "            mat['indices'][:], \n",
    "            mat['indptr'][:]\n",
    "        ), \n",
    "        shape=(mat['shape'][1],mat['shape'][0])\n",
    "    )\n",
    "    gene_dict = dict(mat['features'])\n",
    "    \n",
    "    barcodes = np.array([x.decode('ascii') + f\"-{str(suffix)}\" for x in mat['barcodes']])\n",
    "    _, idx = np.unique(barcodes, return_index=True)\n",
    "    x = x[idx, :]\n",
    "    barcodes = barcodes[idx]\n",
    "    \n",
    "    return x, barcodes, gene_dict\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1a45dff-1b72-44d2-ae83-1013f124cefd",
   "metadata": {},
   "source": [
    "### Create AnnData from filtered samples with ambient mRNA removed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "id": "7998793b-0aaa-4872-9f14-da46f0b7441f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of filtered data files found: 18\n"
     ]
    }
   ],
   "source": [
    "filtered_fns = [fn for fn in os.listdir(base_dir) if \"_filtered.h5\" in fn]\n",
    "print(f\"Number of filtered data files found: {len(filtered_fns)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "id": "5fdf5a2c-f459-4000-99b3-6127901a547b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "WTC_0min_PM_filtered.h5 WTC_PM [13] PM 0\n",
      "WTG_1hr_filtered.h5 WTG_PM [15] PM 1\n",
      "WTD_6hr_Post_filtered.h5 WTD_PM [17] PM 6\n",
      "WTB_0min_PM_filtered.h5 WTB_PM [7] PM 0\n",
      "WTF_1hr_filtered.h5 WTF_PM [5] PM 1\n",
      "WTB_0min_Living_PFC_filtered.h5 WTB_Living [6] Living 0\n",
      "913_0min_filtered.h5 913_Living [8] Living 0\n",
      "WTG_0min_filtered.h5 WTG_Living [14] Living 0\n",
      "WTE_6hr_Post_filtered.h5 WTE_PM [11] PM 6\n",
      "WTA_0min_Living_PFC_filtered.h5 WTA_Living [0] Living 0\n",
      "WTE_0min_Living_filtered.h5 WTE_Living [10] Living 0\n",
      "WTD_0min_Living_filtered.h5 WTD_Living [16] Living 0\n",
      "WTC_0min_Living_filtered.h5 WTC_Living [12] Living 0\n",
      "WTA_0min_PM_filtered.h5 WTA_PM [1] PM 0\n",
      "912_0min_filtered.h5 912_Living [2] Living 0\n",
      "912_5min_filtered.h5 912_PM [3] PM 5\n",
      "WTF_0min_filtered.h5 WTF_Living [4] Living 0\n",
      "913_5min_filtered.h5 913_PM [9] PM 5\n",
      "Variables added. Number of genes remaining: 56481\n",
      "Saved /sc/arion/work/massen06/living_brain1/anndata/mouse_living_brain_filtered.h5ad\n"
     ]
    }
   ],
   "source": [
    "x = None\n",
    "cell_counts = []\n",
    "\n",
    "for n, fn in enumerate(filtered_fns):\n",
    "\n",
    "    # Load filtered data (ambient mRNA already removed)\n",
    "    x_temp, barcodes_temp, gene_dict = readCellbenderH5(os.path.join(base_dir, fn), n)\n",
    "    n_cells = len(barcodes_temp)\n",
    "    cell_counts.append(x_temp.shape[0])\n",
    "    \n",
    "    \n",
    "    # Find the matching donor ID\n",
    "    idx = fn.find(\"filtered\")\n",
    "    sample_id = fn[: idx - 1] \n",
    "    \n",
    "    i0 = fn.find(\"_\")\n",
    "    donor = fn[: i0]\n",
    "    pmi = int(\"\".join([s for s in fn[i0 :-2] if s.isdigit()]))\n",
    "    living = \"PM\" if \"PM\" in fn or \"Post\" in fn or pmi > 0 else \"Living\"\n",
    "    \n",
    "    unique_id = donor + \"_\" + living \n",
    "    \n",
    "    donor_idx = [i for i, donor_id in enumerate(donor_info.unique_id) if unique_id == donor_id]\n",
    "    assert len(donor_idx) <= 1, \"Multiple matches for sample ID\"\n",
    "    print(fn, unique_id, donor_idx, living, pmi)\n",
    "\n",
    "        \n",
    "    if x is None:\n",
    "        x = copy.deepcopy(x_temp)\n",
    "        obs = {\"barcodes\": list(barcodes_temp)}\n",
    "        for k, v in donor_info.iloc[donor_idx[0]].items():\n",
    "            obs[k] = [v] * n_cells\n",
    "\n",
    "    else:        \n",
    "        x = sparse.vstack([x, x_temp])\n",
    "        for k, v in donor_info.iloc[donor_idx[0]].items():\n",
    "            obs[k] += [v] * n_cells\n",
    "        obs[\"barcodes\"] +=  list(barcodes_temp)\n",
    "\n",
    "        \n",
    "adata = sc.AnnData(x)\n",
    "for k, v in obs.items():\n",
    "    adata.obs[k] = v\n",
    "    \n",
    "adata.var = pd.DataFrame({\"gene_ids\": [g.decode('ascii') for g in gene_dict[\"id\"]]})\n",
    "#adata.var.gene_ids = [g.decode('ascii') for g in gene_dict[\"id\"]]\n",
    "    \n",
    "\n",
    "adata = link_gene_info(adata, gene_info)\n",
    "adata.var.index = adata.var.gene_name\n",
    "\n",
    "adata.var.gene_chrom = adata.var.gene_chrom.astype('str')\n",
    "adata.var_names_make_unique()\n",
    "\n",
    "anndata_fn = os.path.join(save_dir, \"mouse_living_brain_filtered.h5ad\")\n",
    "adata.write(anndata_fn)\n",
    "print(f\"Saved {anndata_fn}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "ce99399d-903d-4047-bb78-849928dc1f9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 80703 × 56481\n",
       "    obs: 'barcodes', 'unique_id', 'SubID', 'pmi', 'living', 'step_2_date', 'cdna_amp_cycle', 'input_ng', 'pcr_cycle'\n",
       "    var: 'gene_name', 'gene_type', 'gene_id', 'gene_chrom', 'gene_start', 'gene_end'"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "abaa0d3b-c0df-4d1b-8d2a-29dfeab124a3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "707 6525 4779.5\n"
     ]
    }
   ],
   "source": [
    "print(np.min(cell_counts), np.max(cell_counts), np.median(cell_counts))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4df10c8e-b573-46f3-a637-b3147498643e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qc_torch",
   "language": "python",
   "name": "qc_torch"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
