{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ddfe775a-8d8a-43d0-a25a-6de438f06cfd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/sc/arion/work/massen06/.conda/envs/qc/bin/python3.8\n",
      "     active environment : qc\n",
      "    active env location : /sc/arion/work/massen06/.conda/envs/qc\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "print(sys.executable)\n",
    "!conda info | grep 'active env'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "1b222304-72ca-4d5b-b389-1f186d3576b8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/sc/arion/work/massen06/.conda/envs/qc/lib/python3.8/site-packages/umap/distances.py:1063: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "/sc/arion/work/massen06/.conda/envs/qc/lib/python3.8/site-packages/umap/distances.py:1071: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "/sc/arion/work/massen06/.conda/envs/qc/lib/python3.8/site-packages/umap/distances.py:1086: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n",
      "/sc/arion/work/massen06/.conda/envs/qc/lib/python3.8/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/sc/arion/work/massen06/.conda/envs/qc/lib/python3.8/site-packages/umap/umap_.py:660: NumbaDeprecationWarning: \u001b[1mThe 'nopython' keyword argument was not supplied to the 'numba.jit' decorator. The implicit default value for this argument is currently False, but it will be changed to True in Numba 0.59.0. See https://numba.readthedocs.io/en/stable/reference/deprecation.html#deprecation-of-object-mode-fall-back-behaviour-when-using-jit for details.\u001b[0m\n",
      "  @numba.jit()\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.9.3 anndata==0.9.1 umap==0.5.3 numpy==1.24.4 scipy==1.10.1 pandas==2.0.3 scikit-learn==1.3.0 statsmodels==0.14.0 python-igraph==0.10.6 louvain==0.8.1 pynndescent==0.5.10\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import copy\n",
    "import pge\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import pegasus as pg\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import matplotlib.pyplot as plt\n",
    "from datetime import datetime"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ae9585e-5645-4b34-ab75-6bf4ff510030",
   "metadata": {},
   "source": [
    "# Mouse Living Brain - Data Loading\n",
    "Nicolas Masse, Donghoon Lee  \n",
    "July 2023"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e3b0ad4-711d-44a9-ac1e-8d042a908b24",
   "metadata": {},
   "source": [
    "### Paths for metadata spreadsheet, gene info, and target path for AnnData"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "5dacdda8-bd66-431a-b7e0-27742d02be31",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_dir = \"/sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/metadata\"\n",
    "base_dir = \"/sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams\"\n",
    "save_dir = \"/sc/arion/work/massen06/living_brain1/anndata\"\n",
    "gene_info_fn = \"/hpc/users/massen06/MM_GRCm39_full.txt\" \n",
    "mouse_mitocarta_fn = \"MouseMitoCarta3.0.csv\" "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc6be895-0d2b-406c-8abd-f031dc5cab91",
   "metadata": {},
   "source": [
    "### Extract metadata from Living_brain_snRNA_template.xlsx\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "294579c9-6577-4fe9-b24b-60b01780d74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "metadata_fn = os.path.join(metadata_dir, \"Living_brain_snRNA_template.xlsx\")\n",
    "metadata = pd.read_excel(metadata_fn, sheet_name=\"Mouse_Master Libraries\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d12d1305-e8c7-441b-9920-40c98ad42c99",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_donor_information(data):\n",
    "    \"\"\"Extract donor name, PMI and living status\"\"\"\n",
    "    \n",
    "    donor_info = {\n",
    "        \"unique_id\": [],\n",
    "        \"SubID\" : [],\n",
    "        \"pmi\": [],\n",
    "        \"living\": [],\n",
    "    }\n",
    "\n",
    "    for d in data.values:\n",
    "        \n",
    "        donor_info[\"unique_id\"].append(d)\n",
    "        \n",
    "        # get donor name\n",
    "        i0 = d.find(\"_\")\n",
    "        donor_info[\"SubID\"].append(d[: i0])\n",
    "        \n",
    "        # get PMI\n",
    "        time = int(\"\".join([s for s in d[i0 :] if s.isdigit()]))\n",
    "        if \"hr\" in d[i0 :]:\n",
    "            time *= 60\n",
    "        donor_info[\"pmi\"].append(time)\n",
    "        \n",
    "        if \"PM\" in d or \"Post\" in d or time > 0:\n",
    "            donor_info[\"living\"].append(False)\n",
    "        else:\n",
    "            donor_info[\"living\"].append(True)\n",
    "        \n",
    "    return donor_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bb835560-fcd8-4461-9671-986b63d93c0b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "              unique_id SubID  pmi  living   Channel  cdna_amp_cycle  \\\n",
      "0   WTA_0min_Living_PFC   WTA    0    True  06022023              11   \n",
      "1           WTA_0min_PM   WTA    0   False  06022023              11   \n",
      "2              912_0min   912    0    True  06022023              11   \n",
      "3              912_5min   912    5   False  06022023              11   \n",
      "4              WTF_0min   WTF    0    True  06022023              11   \n",
      "5               WTF_1hr   WTF   60   False  06022023              11   \n",
      "6   WTB_0min_Living_PFC   WTB    0    True  06062023              12   \n",
      "7           WTB_0min_PM   WTB    0   False  06062023              12   \n",
      "8              913_0min   913    0    True  06062023              12   \n",
      "9              913_5min   913    5   False  06062023              12   \n",
      "10      WTE_0min_Living   WTE    0    True  06062023              12   \n",
      "11         WTE_6hr_Post   WTE  360   False  06062023              12   \n",
      "12      WTC_0min_Living   WTC    0    True  06072023              12   \n",
      "13          WTC_0min_PM   WTC    0   False  06072023              12   \n",
      "14             WTG_0min   WTG    0    True  06072023              12   \n",
      "15              WTG_1hr   WTG   60   False  06072023              12   \n",
      "16      WTD_0min_Living   WTD    0    True  06072023              12   \n",
      "17         WTD_6hr_Post   WTD  360   False  06072023              12   \n",
      "\n",
      "    input_ng  pcr_cycle  \n",
      "0      139.6         12  \n",
      "1      174.0         12  \n",
      "2       40.8         13  \n",
      "3      146.8         12  \n",
      "4       31.4         13  \n",
      "5      140.4         12  \n",
      "6       28.0         14  \n",
      "7       40.0         13  \n",
      "8       21.4         14  \n",
      "9       47.6         13  \n",
      "10      16.0         14  \n",
      "11      50.0         13  \n",
      "12     460.0         10  \n",
      "13     384.0         10  \n",
      "14     164.0         12  \n",
      "15     404.0         10  \n",
      "16      34.2         13  \n",
      "17     296.0         11  \n"
     ]
    }
   ],
   "source": [
    "for k in metadata.keys():\n",
    "    \n",
    "    column_name = metadata[k].iloc[0]\n",
    "\n",
    "    if column_name == \"unique sample ID\":\n",
    "        donor_info = extract_donor_information(metadata[k].iloc[1 :])\n",
    "        \n",
    "    elif column_name == \"Step 2 date\":\n",
    "        dates = metadata[k].iloc[1 :]\n",
    "        dates = [d.strftime(\"%m%d%Y\") for d in dates]\n",
    "        donor_info[\"Channel\"] = dates\n",
    "        \n",
    "    elif column_name == \"cDNA Amp cycle #s\":\n",
    "        cycles = metadata[k].iloc[1 :]\n",
    "        cycles = [int(c) for c in cycles]\n",
    "        donor_info[\"cdna_amp_cycle\"] = cycles\n",
    "        \n",
    "    elif column_name == \"input (ng)\":\n",
    "        inputs = metadata[k].iloc[1 :]\n",
    "        inputs = [round(float(i), 1) for i in inputs]\n",
    "        donor_info[\"input_ng\"] = inputs\n",
    "        \n",
    "    elif column_name == \"Indexing PCR cycle #\":\n",
    "        cycle = metadata[k].iloc[1 :]\n",
    "        cycle = [int(c) for c in cycle]\n",
    "        donor_info[\"pcr_cycle\"] = cycle\n",
    "        \n",
    "donor_info = pd.DataFrame(data = donor_info)\n",
    "\n",
    "print(donor_info)  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e8bc0c-0324-4684-8ac7-37ab18f012e0",
   "metadata": {},
   "source": [
    "### Load gene info from Biomart"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "efc2a921-f16c-48c2-8943-7107d0a513d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_176025/2442110492.py:1: DtypeWarning: Columns (9) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  gene_info = pd.read_csv(gene_info_fn, sep='\\t')\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>index</th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene start (bp)</th>\n",
       "      <th>Gene end (bp)</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>Gene type</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Chromosome/scaffold name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>ENSMUSG00000064336.1</td>\n",
       "      <td>ENSMUST00000082387</td>\n",
       "      <td>ENSMUST00000082387.1</td>\n",
       "      <td>1</td>\n",
       "      <td>68</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA phenylalanine [So...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>ENSMUSG00000064337.1</td>\n",
       "      <td>ENSMUST00000082388</td>\n",
       "      <td>ENSMUST00000082388.1</td>\n",
       "      <td>70</td>\n",
       "      <td>1024</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:MGI S...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>ENSMUSG00000064338.1</td>\n",
       "      <td>ENSMUST00000082389</td>\n",
       "      <td>ENSMUST00000082389.1</td>\n",
       "      <td>1025</td>\n",
       "      <td>1093</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA valine [Source:MG...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>ENSMUSG00000064339.1</td>\n",
       "      <td>ENSMUST00000082390</td>\n",
       "      <td>ENSMUST00000082390.1</td>\n",
       "      <td>1094</td>\n",
       "      <td>2675</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:MGI S...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>ENSMUSG00000064340.1</td>\n",
       "      <td>ENSMUST00000082391</td>\n",
       "      <td>ENSMUST00000082391.1</td>\n",
       "      <td>2676</td>\n",
       "      <td>2750</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>mitochondrially encoded tRNA leucine 1 [Source...</td>\n",
       "      <td>MT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56477</th>\n",
       "      <td>149518</td>\n",
       "      <td>ENSMUSG00000054766</td>\n",
       "      <td>ENSMUSG00000054766.14</td>\n",
       "      <td>ENSMUST00000134364</td>\n",
       "      <td>ENSMUST00000134364.8</td>\n",
       "      <td>29947390</td>\n",
       "      <td>29962589</td>\n",
       "      <td>Set</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>SET nuclear oncogene [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56478</th>\n",
       "      <td>149525</td>\n",
       "      <td>ENSMUSG00000026785</td>\n",
       "      <td>ENSMUSG00000026785.15</td>\n",
       "      <td>ENSMUST00000125346</td>\n",
       "      <td>ENSMUST00000125346.8</td>\n",
       "      <td>29967696</td>\n",
       "      <td>29981034</td>\n",
       "      <td>Pkn3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>protein kinase N3 [Source:MGI Symbol;Acc:MGI:2...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56479</th>\n",
       "      <td>149533</td>\n",
       "      <td>ENSMUSG00000015335</td>\n",
       "      <td>ENSMUSG00000015335.17</td>\n",
       "      <td>ENSMUST00000148717</td>\n",
       "      <td>ENSMUST00000148717.8</td>\n",
       "      <td>29980956</td>\n",
       "      <td>29983660</td>\n",
       "      <td>Zdhhc12</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>zinc finger, DHHC domain containing 12 [Source...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56480</th>\n",
       "      <td>149537</td>\n",
       "      <td>ENSMUSG00000039686</td>\n",
       "      <td>ENSMUSG00000039686.15</td>\n",
       "      <td>ENSMUST00000044751</td>\n",
       "      <td>ENSMUST00000044751.14</td>\n",
       "      <td>29987295</td>\n",
       "      <td>30014597</td>\n",
       "      <td>Zer1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>zyg-11 related, cell cycle regulator [Source:M...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56481</th>\n",
       "      <td>149546</td>\n",
       "      <td>ENSMUSG00000100747</td>\n",
       "      <td>ENSMUSG00000100747.6</td>\n",
       "      <td>ENSMUST00000187343</td>\n",
       "      <td>ENSMUST00000187343.2</td>\n",
       "      <td>30126807</td>\n",
       "      <td>30128235</td>\n",
       "      <td>1700084E18Rik</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>RIKEN cDNA 1700084E18 gene [Source:MGI Symbol;...</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>56482 rows Ã— 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        index      Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0           0  ENSMUSG00000064336   ENSMUSG00000064336.1   ENSMUST00000082387   \n",
       "1           1  ENSMUSG00000064337   ENSMUSG00000064337.1   ENSMUST00000082388   \n",
       "2           2  ENSMUSG00000064338   ENSMUSG00000064338.1   ENSMUST00000082389   \n",
       "3           3  ENSMUSG00000064339   ENSMUSG00000064339.1   ENSMUST00000082390   \n",
       "4           4  ENSMUSG00000064340   ENSMUSG00000064340.1   ENSMUST00000082391   \n",
       "...       ...                 ...                    ...                  ...   \n",
       "56477  149518  ENSMUSG00000054766  ENSMUSG00000054766.14   ENSMUST00000134364   \n",
       "56478  149525  ENSMUSG00000026785  ENSMUSG00000026785.15   ENSMUST00000125346   \n",
       "56479  149533  ENSMUSG00000015335  ENSMUSG00000015335.17   ENSMUST00000148717   \n",
       "56480  149537  ENSMUSG00000039686  ENSMUSG00000039686.15   ENSMUST00000044751   \n",
       "56481  149546  ENSMUSG00000100747   ENSMUSG00000100747.6   ENSMUST00000187343   \n",
       "\n",
       "      Transcript stable ID version  Gene start (bp)  Gene end (bp)  \\\n",
       "0             ENSMUST00000082387.1                1             68   \n",
       "1             ENSMUST00000082388.1               70           1024   \n",
       "2             ENSMUST00000082389.1             1025           1093   \n",
       "3             ENSMUST00000082390.1             1094           2675   \n",
       "4             ENSMUST00000082391.1             2676           2750   \n",
       "...                            ...              ...            ...   \n",
       "56477         ENSMUST00000134364.8         29947390       29962589   \n",
       "56478         ENSMUST00000125346.8         29967696       29981034   \n",
       "56479         ENSMUST00000148717.8         29980956       29983660   \n",
       "56480        ENSMUST00000044751.14         29987295       30014597   \n",
       "56481         ENSMUST00000187343.2         30126807       30128235   \n",
       "\n",
       "           Gene name       Gene type  \\\n",
       "0              mt-Tf         Mt_tRNA   \n",
       "1            mt-Rnr1         Mt_rRNA   \n",
       "2              mt-Tv         Mt_tRNA   \n",
       "3            mt-Rnr2         Mt_rRNA   \n",
       "4             mt-Tl1         Mt_tRNA   \n",
       "...              ...             ...   \n",
       "56477            Set  protein_coding   \n",
       "56478           Pkn3  protein_coding   \n",
       "56479        Zdhhc12  protein_coding   \n",
       "56480           Zer1  protein_coding   \n",
       "56481  1700084E18Rik          lncRNA   \n",
       "\n",
       "                                        Gene description  \\\n",
       "0      mitochondrially encoded tRNA phenylalanine [So...   \n",
       "1      mitochondrially encoded 12S rRNA [Source:MGI S...   \n",
       "2      mitochondrially encoded tRNA valine [Source:MG...   \n",
       "3      mitochondrially encoded 16S rRNA [Source:MGI S...   \n",
       "4      mitochondrially encoded tRNA leucine 1 [Source...   \n",
       "...                                                  ...   \n",
       "56477  SET nuclear oncogene [Source:MGI Symbol;Acc:MG...   \n",
       "56478  protein kinase N3 [Source:MGI Symbol;Acc:MGI:2...   \n",
       "56479  zinc finger, DHHC domain containing 12 [Source...   \n",
       "56480  zyg-11 related, cell cycle regulator [Source:M...   \n",
       "56481  RIKEN cDNA 1700084E18 gene [Source:MGI Symbol;...   \n",
       "\n",
       "      Chromosome/scaffold name  \n",
       "0                           MT  \n",
       "1                           MT  \n",
       "2                           MT  \n",
       "3                           MT  \n",
       "4                           MT  \n",
       "...                        ...  \n",
       "56477                        2  \n",
       "56478                        2  \n",
       "56479                        2  \n",
       "56480                        2  \n",
       "56481                        2  \n",
       "\n",
       "[56482 rows x 11 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_info = pd.read_csv(gene_info_fn, sep='\\t')\n",
    "gene_info.drop_duplicates(subset = \"Gene name\", inplace=True)\n",
    "gene_info.reset_index(inplace=True)\n",
    "gene_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "5cb1cadf-a7e6-48e9-b744-dbf282a05c09",
   "metadata": {},
   "outputs": [],
   "source": [
    "def link_gene_info(adata, gene_info, verbose=False):\n",
    "    \"\"\"Add gene info to AnnData\"\"\"\n",
    "    \n",
    "    adata.var = adata.var.merge(gene_info, left_on=\"gene_ids\", right_on='Gene stable ID', how=\"outer\")\n",
    "\n",
    "    good_genes_idx = [i for i, name in enumerate(adata.var[\"Gene name\"].values) if isinstance(name, str)]\n",
    "       \n",
    "    new_adata = ad.AnnData(adata.X[:, good_genes_idx])\n",
    "    new_adata.obs = adata.obs\n",
    "    new_adata.var[\"gene_name\"] = [v for n, v in enumerate(adata.var[\"Gene name\"].values) if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_type\"] = [v for n, v in enumerate(adata.var[\"Gene type\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_id\"] = [v for n, v in enumerate(adata.var[\"Gene stable ID\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_chrom\"] = [v for n, v in enumerate(adata.var[\"Chromosome/scaffold name\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_start\"] = [v for n, v in enumerate(adata.var[\"Gene start (bp)\"].values)  if n in good_genes_idx]\n",
    "    new_adata.var[\"gene_end\"] = [v for n, v in enumerate(adata.var[\"Gene end (bp)\"].values)  if n in good_genes_idx]\n",
    "    \n",
    "    print(f\"Variables added. Number of genes remaining: {new_adata.X.shape[1]}\")\n",
    "    \n",
    "    return new_adata"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1a45dff-1b72-44d2-ae83-1013f124cefd",
   "metadata": {},
   "source": [
    "### Create AnnData from moise samples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7998793b-0aaa-4872-9f14-da46f0b7441f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of target directories found: 18\n"
     ]
    }
   ],
   "source": [
    "target_dirs = [fn for fn in os.listdir(base_dir) if \"Solo.out\" in fn]\n",
    "print(f\"Number of target directories found: {len(target_dirs)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "aa792ce6-5099-4310-b3da-216362019547",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_912_0min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTE_0min_Living_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTF_0min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTD_6hr_Post_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTD_0min_Living_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTG_0min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTF_1hr_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTC_0min_PM_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_913_5min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTE_6hr_Post_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_912_5min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTG_1hr_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTB_0min_PM_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTA_0min_PM_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTA_0min_Living_PFC_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTB_0min_Living_PFC_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_913_0min_Solo.out/GeneFull/filtered\n",
      "Creating anndata from /sc/arion/projects/CommonMind/single-cell-neurogenomics/LIVING_BRAIN/pre_processed_files/STARsolo_bams/LB_WTC_0min_Living_Solo.out/GeneFull/filtered\n",
      "Variables added. Number of genes remaining: 56481\n",
      "Saved /sc/arion/work/massen06/living_brain1/anndata/mouse_living_brain.h5ad\n"
     ]
    }
   ],
   "source": [
    "adata = None\n",
    "\n",
    "for n, t in enumerate(target_dirs):\n",
    "\n",
    "    # Load 10x data\n",
    "    data_path = os.path.join(base_dir, t, \"GeneFull\", \"filtered\")\n",
    "    print(f\"Creating anndata from {data_path}\")\n",
    "    adata_temp = sc.read_10x_mtx(data_path)\n",
    "    \n",
    "    # Find the matching donor ID\n",
    "    idx = t.find(\"Solo\")\n",
    "    sample_id = t[3: idx - 1] # index starts at 3 to remove prefix \"LB_\"\n",
    "    donor_idx = [i for i, donor_id in enumerate(donor_info.unique_id) if sample_id in donor_id]\n",
    "    assert len(donor_idx) <= 1, \"Multiple matches for sample ID\"\n",
    "    \n",
    "    for k, v in donor_info.iloc[donor_idx[0]].items():\n",
    "        adata_temp.obs[k] = v\n",
    "    if adata is None:\n",
    "        adata = copy.deepcopy(adata_temp)\n",
    "    else:        \n",
    "        # adata = ad.concat([adata, adata_temp], axis=0, merge=\"same\")   \n",
    "        adata = adata.concatenate(adata_temp)\n",
    "\n",
    "for k, v in adata_temp.var.items():\n",
    "    adata.var[k] = v\n",
    "\n",
    "adata = link_gene_info(adata, gene_info)\n",
    "adata.var.index = adata.var.gene_name\n",
    "\n",
    "adata.var.gene_chrom = adata.var.gene_chrom.astype('str')\n",
    "\n",
    "anndata_fn = os.path.join(save_dir, \"mouse_living_brain.h5ad\")\n",
    "adata.write(anndata_fn)\n",
    "print(f\"Saved {anndata_fn}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ef7f254d-4acc-4bed-aa34-11219dbb7dcf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "qc",
   "language": "python",
   "name": "qc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
